### ISSUE TITLE
Seeing slab_reassign_evictions_nomem when items move to SSD

### ISSUE THREAD

**Issue by sergmour** [memcached/memcached#794]
We are actively using ExtStore feature and we see evictions (slab_reassign_evictions_nomem) during long running data ingestion jobs.
In many of our deployments we hit the limit of available RAM. These are big VMs with 112G of memory and 800G of SSD. Our observation is that during long running ingestion jobs items get shuffled causing some items in RAM to be moved to SSD. We have enough free space on SSD and we expect no evictions - items to be simply moved from RAM onto SSD. However, at some point we start seeing evictions (coinciding with slab_reassign_evictions_nomem metric). With slab_automove_freeratio defaulted to 1%, the rate of evictions becomes very high (thousands per sec). When we set slab_automove_freeratio to 20%-25%, evictions get postponed and happen occasionally with a considerably lower rate (tens per sec).
It might be a speculation at this point, but we think what happens is that when an item needs to be offloaded onto SSD, its metadata gets moved from a higher to a lower slab. When the lower slab does not have space to grow, it simply evicts one of the other LRU keys (apparently, another metadata for another SSD-offloaded item).
No matter how high we set slab_automove_freeratio (up to 40%), there is still a point when evictions start happening and our customers complain.
Any thoughts/suggestions on how to eliminate or mitigate these evictions would be greatly appreciated.

**Comment by dormando** [comment]
Have you tried manually setting `ext_item_age` to some low level? That can greatly increase the aggressiveness of the flushing and help avoid slab page move scenarios that can lead to `_nomem` evictions.

I've had this thought for a while though, see here: https://github.com/memcached/memcached/issues/541#issuecomment-619457044
It's a few significant steps but entirely removes `_nomem` as a possibility and would simplify a few related systems.

If your co's up for a work contract I can see about prioritizing that change higher or doing more detailed analysis on what's going wrong.

**Comment by sergmour** [comment]
Thanks for the ext_item_age hint, @dormando! Will try it. I believe we should be able to set it without restart required. #541 looks promising but involved. I doubt about the contract, but will ask.

**Comment by sergmour** [comment]
@dormando, setting ext_item_age does help with the heavy-ssd-writes case, thanks again for the hint.

**Comment by dormando** [comment]
Great to hear! I agree it does _feel_ lame. `ext_item_age` was the original flush method while I was developing extstore, and I only left it in for compatibility; but then other folks ended up using it. I would be curious to know if you can actually measure a difference in read performance, and how significant that is if so.

Also for what it's worth, if you reduce the size of the cache daemon a bit (or if it hasn't allocated all of memory) then the OS gets more filesystem buffer space, which is used by extstore presently. For items recently flushed to SSD you'd still get them pulled back without hitting disk, just a slightly slower path internally to memcached.

Another thing is some folks will set `ext_item_age` while doing a bulk/fast load, then disable it once the bulk load is done. Since bulk loads have no information on the value of what should actually _be_ in memory, letting that resolve once the bulk load is done can help (via tuning the recache-rate if necessary).

**Comment by sergmour** [comment]
Can't give you much more trusted data at this point. We are a kind of "service-provider" team within a very large org providing memcached deployments with mcrouter proxies between the clients and the shards. We have limited visibility on customer flows.
Yesterday we set the age to 15 sec on one of the clusters and the RAM utilization went down to almost 0. There was a performance test and it looks like it held pretty well. The average latency on Mcrouter side increased twice but still held at half-msec on small payloads. And the max was still good enough. This includes a network hop from Mcrouter to memcached. We run on VMs and the cloud provider is capping the disk IO, though, the latency.

**Comment by dormando** [comment]
Understood. Also keep in mind the slab rebalancer algorithm can be run externally to the daemon (see scripts/memcached-automove*). You might be able to customize something that works better, possibly even by automatically adjusting `ext_item_age` when the global page pool gets too low (on top of the existing algo). There are some options that shouldn't be too painful while you wait.

RE: Proxy. We're still a ways off from being production ready, but I'm just about at the point of designing the route API. If you folks are interested in helping, taking a look would be much appreciated. There's a long list of further work but not everyone will need everything to go to production (or just to test), so any hints as to what's most important for you will help a lot.

Once I get through a pass on the API and more confidence in stability I can start documenting / blogging about it, which should make it easier to get into. You can get pretty far starting from the example configuration noted in the PR though.

**Comment by dormando** [comment]
https://github.com/memcached/memcached/issues/796 for proxy discussion.

**Comment by dormando** [comment]
https://github.com/memcached/memcached/pull/923 - this change might alleviate the need for `ext_item_age = 0` in this use case, ahead of the planned larger rewrite that I've not had time for.

**Comment by dormando** [comment]
This has actually been fixed in several ways over the last year. The `_nomem` case has been completely removed and flushing speeds generally improved. Running the latest version should allow you to use all of the memory without setting `item_age=0`
